crates/ra_hir_def/src/item_tree.rs

```diff
@@ -500,13 +500,19 @@ pub struct Function {
     pub name: Name,
     pub visibility: RawVisibilityId,
     pub generic_params: GenericParamsId,
-    pub has_self_param: bool,
+    pub self_param: Option<SelfParam>,
     pub is_unsafe: bool,
     pub params: Box<[TypeRef]>,
     pub ret_type: TypeRef,
     pub ast_id: FileAstId<ast::FnDef>,
 }
 
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub struct SelfParam {
+    pub is_ref: bool,
+    pub is_mut: bool,
```

Comment:
This information is already contained in the TypeRef for the self param, I don't think it's a good idea to duplicate it

Reply:
So if `has_self_param` is `true`, then I can treat the first param as if it was `self`, right?

Reply:
Yes, exactly.

Reply:
That works. Done now